sklearn/model_selection/_split.py

```diff
@@ -601,17 +602,29 @@ def __init__(self, n_splits='warn', shuffle=False, random_state=None):
 
     def _make_test_folds(self, X, y=None):
         rng = self.random_state
-        y = np.asarray(y)
-        type_of_target_y = type_of_target(y)
+
+        # For compatibility with GridSearchCV (which accepts y to be a sparse
+        # matrix), if y is sparse we convert it to a dense matrix. An error is
+        # raised if y does not have the right shape, to avoid creating a large
+        # dense matrix (whose shape will raise an error in ``column_or_1d``).
+        if sp.issparse(y):
+            shape = y.shape
+            if len(shape) > 2 or (len(shape) == 2 and shape[1] != 1):
+                raise ValueError("bad input shape {0}".format(shape))
+            else:
+                _y = y.toarray()
```

Comment:
If the data is sparse, we must assume it's a bad idea to automatically convert it to sparse. (I.e. it should create memory problems.) I would rather reject sparse y from Stratified* than densify it.

Reply:
Converting y to a dense array was a workaround to allow for the case where one does binary classification with a sparse y (here, y should have shape (n_samples,) or (n_samples, 1)). In this case, converting it to a dense array should not create memory problems... If we reject y from Stratified*, we do not allow for (imbalanced) binary classif with sparse y. Are you sure we should reject it in Stratified*?

Reply:
I'm okay with that at least for now. I don't see why a sparse matrix should be especially useful for a single column

Though I suppose I can accept the current behaviour as long as we are sure it's a single column...

Reply:
@GaelVaroquaux OK to reject a 1D sparse y in Stratified* (at least for now) or do you prefer to keep it?